core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java

```diff
@@ -1556,5 +1561,230 @@ private void setInterruptFlagInternal(AtomicBoolean flag) {
     public void setCacheProvider(CacheProvider cacheProvider) {
       reader.setCacheProvider(cacheProvider);
     }
+
+    @Override
+    public void reset() {
+      clear();
+    }
+  }
+
+  public interface RFileSKVIterator extends FileSKVIterator {
+    FileSKVIterator getIndex() throws IOException;
+
+    void reset();
+  }
+
+  static abstract class FencedFileSKVIterator implements FileSKVIterator {
+
+    private final FileSKVIterator reader;
+    protected final Range fence;
+    private final Key fencedStartKey;
+    private final Key fencedEndKey;
+
+    public FencedFileSKVIterator(FileSKVIterator reader, Range fence) {
+      this.reader = Objects.requireNonNull(reader);
+      this.fence = Objects.requireNonNull(fence);
+      this.fencedStartKey = fence.getStartKey();
+      this.fencedEndKey = getEndKey(fence.getEndKey());
```

Comment:
I would suggest moving this computation of the fencedEndKEy to the getLastKey() method.  Reason is that not every scan will need this.  Only things like splitting a tablet will use it.  So I suspect its being frequently computed but not frequently used.  Maybe the fencedEndKey instance var could be dropped.

```suggestion
```

Reply:
That's a good point, do you think it makes sense to lazy cache the value after computation for repeat calls to `getLastKey()` on the same reader instance?

I haven't looked in detail but after a quick glance it appears that in the split code where `getLastKey()` is used it will just be called once but I'm wondering if there could be a use case in the future (or current that I missed) where the method could be called multiple times on the same reader so it would be worthwhile to avoid having to compute and copy the bytes each time.

Reply:
@keith-turner - another option I thought of was just caching using Guava memoize which we've done elsewhere. I created a separate branch/commit as a quick example of how this would look and wanted to see what you thought: https://github.com/cshannon/accumulo/commit/d871e466ac90a02dfd78fdbda90ca4a1b0d8cfa7

Reply:
> another option I thought of was just caching using Guava memoize which we've done elsewhere. 

I like that solution, multiple benefits w/o any downside.